src/visitors/cssNamespace.js

```diff
@@ -1,21 +1,49 @@
 import * as t from 'babel-types';
 import { isStyled } from 'babel-plugin-styled-components/lib/utils/detectors';
 
+const containsInitialSelfReference = css =>
+  /^(([^\{]([\w-_,:+\s]+))*\&\s*([^\{]([\w-_,:+\s]+))*)+[\{]/g.test(css.trim());
+
 const getCssNamespace = state => {
-  const { cssNamespace } = state.opts;
-  if (!cssNamespace) {
-    return '&&';
+  const { cssNamespace, postProcessCssNamespace } = state.opts;
+  if (!cssNamespace && !postProcessCssNamespace) {
+    return { cssNamespace: '&&', hasNamespaceSelfReference: true };
   }
 
-  if (Array.isArray(cssNamespace)) {
-    return `.${cssNamespace.join(' .')} &`;
+  if (cssNamespace) {
+    const wrapperClass = `.${[].concat(cssNamespace).join(' .')} &`;
+    const hasNamespaceSelfReference = containsInitialSelfReference(
+      `${wrapperClass} {`
+    );
+    return { cssNamespace: wrapperClass, hasNamespaceSelfReference };
   }
 
-  return `.${cssNamespace} &`;
+  if (postProcessCssNamespace) {
+    const wrapperClass = `${[].concat(postProcessCssNamespace).join(', ')}`;
+    const hasNamespaceSelfReference = containsInitialSelfReference(
+      `${wrapperClass} {`
+    );
+
+    return { cssNamespace: wrapperClass, hasNamespaceSelfReference };
+  }
 };
 
+const wrapCssDefinition = (
+  cssNamespace,
+  css,
+  isSelfReferenced,
+  shouldLeaveScopeOpen
+) =>
+  isSelfReferenced
+    ? `${cssNamespace} {${css}${shouldLeaveScopeOpen ? '' : '}'}`
+    : `${cssNamespace} {&{${css}${shouldLeaveScopeOpen ? '' : '}}'}`;
+
+const NO_STRIP = false;
+const STRIP_END_CURLY_BRACKETS = true;
```

Comment:
I'm a fan of this pattern to make booleans as arguments a bit more readable. Very nice!